core/consensus/src/adapter.rs

```diff
@@ -133,18 +139,76 @@ where
         let epoch = self.storage.get_epoch_by_epoch_id(epoch_id).await?;
         Ok(epoch.header.validators)
     }
+
+    async fn get_current_epoch_id(&self, _ctx: Context) -> ProtocolResult<u64> {
+        let res = self.storage.get_latest_proof().await?;
+        Ok(res.epoch_id)
+    }
+
+    async fn pull_epoch(&self, ctx: Context, epoch_id: u64, end: &str) -> ProtocolResult<Epoch> {
+        let msg = FixedEpochID::new(epoch_id);
+        let res = self
+            .rpc
+            .call::<ConsensusRpcRequest, ConsensusRpcResponse>(
+                ctx,
+                end,
+                ConsensusRpcRequest::PullEpochs(msg),
```

Comment:
Why not use `protobuf`?  There is no requirement for serialization consistency in a network.